### ISSUE TITLE
`functools.partial` used on a function with TypeVar loses information

### ISSUE THREAD

**Issue by jack1142** [microsoft/pyright#2220]
**Describe the bug**
Using `functools.partial` on a function accepting a TypeVar seems to lose that type information.

**To Reproduce**
1. Copy code from below
2. Run it, you should see:
```
<class '__main__.Y'>
<class '__main__.X'>
<__main__.X object at 0x000002ED193B39A0>
123
<class '__main__.Y'>
<__main__.Y object at 0x000002ED193B39A0>
123
```
3. Type check it with pyright, as part of the output you should see:
```
deco_returning_type_back.py:39:7 - error: Object of type "object*" is not callable (reportGeneralTypeIssues)
deco_returning_type_back.py:40:7 - error: Object of type "object*" is not callable (reportGeneralTypeIssues)
deco_returning_type_back.py:43:17 - info: Type of "X" is "Type[X]"
deco_returning_type_back.py:44:17 - info: Type of "X(123)" is "X"
deco_returning_type_back.py:45:17 - info: Type of "X(123).method()" is "str"
deco_returning_type_back.py:47:17 - info: Type of "Y" is "T@deco"
deco_returning_type_back.py:48:17 - error: Object of type "object*" is not callable (reportGeneralTypeIssues)
deco_returning_type_back.py:48:17 - info: Type of "Y(123)" is "Unknown"
deco_returning_type_back.py:49:17 - error: Object of type "object*" is not callable (reportGeneralTypeIssues)
deco_returning_type_back.py:49:17 - info: Type of "Y(123).method()" is "Unknown"
```

**Expected behavior**
I expected pyright to keep all of `Y`'s type information correctly:
```
deco_returning_type_back.py:43:17 - info: Type of "X" is "Type[X]"
deco_returning_type_back.py:44:17 - info: Type of "X(123)" is "X"
deco_returning_type_back.py:45:17 - info: Type of "X(123).method()" is "str"
deco_returning_type_back.py:47:17 - info: Type of "Y" is "Type[Y]"
deco_returning_type_back.py:48:17 - info: Type of "Y(123)" is "Y"
deco_returning_type_back.py:49:17 - info: Type of "Y(123).method()" is "str"
```

**Screenshots or Code**
```python
from functools import partial
from typing import TYPE_CHECKING, TypeVar

T = TypeVar("T")

def deco(obj: T, *, print_obj: bool = False) -> T:
    if print_obj:
        print(obj)
    return obj

better_deco = partial(deco, print_obj=True)

@deco
class X:
    def __init__(self, a: int) -> None:
        self.a = a

    def method(self) -> str:
        return str(self.a)

@better_deco
class Y:
    def __init__(self, a: int) -> None:
        self.a = a

    def method(self) -> str:
        return str(self.a)

print(X)
print(X(123))
print(X(123).method())

print(Y)
print(Y(123))
print(Y(123).method())

if TYPE_CHECKING:
    reveal_type(X)
    reveal_type(X(123))
    reveal_type(X(123).method())

    reveal_type(Y)
    reveal_type(Y(123))
    reveal_type(Y(123).method())
```

**VS Code extension or command-line**
Running pyright 1.1.163 as a command-line tool.

**Additional context**
This appears to be working just fine on mypy:
```
deco_returning_type_back.py:43: note: Revealed type is "def (a: builtins.int) -> deco_returning_type_back.X"
deco_returning_type_back.py:44: note: Revealed type is "deco_returning_type_back.X"
deco_returning_type_back.py:45: note: Revealed type is "builtins.str"
deco_returning_type_back.py:47: note: Revealed type is "def (a: builtins.int) -> deco_returning_type_back.Y"
deco_returning_type_back.py:48: note: Revealed type is "deco_returning_type_back.Y"
deco_returning_type_back.py:49: note: Revealed type is "builtins.str"
```

**Comment by jack1142** [comment]
I ended up using a function returning function instead which is inferred correctly by both mypy and pyright + it's a bit more ergnomic in my actual use case anyway:
```py
def get_deco(print_obj: bool = False) -> Callable[[T], T]:
    def deco(obj: T) -> T:
        if print_obj:
            print(obj)
        return obj
    return obj
```

I'm keeping this issue open though as to me, it still does seem like a valid bug.

**Comment by erictraut** [comment]
Here's what's happening.

The class `functools.partial` is generic, and the type argument is derived from the return type of the function that you pass to the `partial` constructor.

```python
class partial(Generic[_T]):
    ...
    def __init__(self, func: Callable[..., _T], *args: Any, **kwargs: Any) -> None: ...
    def __call__(self, *args: Any, **kwargs: Any) -> _T: ...
```

In your case, you're passing a function `deco`, and its return type is an unresolved type variable `T@deco`. Each use of a type variable has a defined scope (see scoping rules in PEP 484 for details), and type variables must be "solved" within their defined scope. In this case, the type variable `T` is scoped to the `deco` function, and it must be resolved in the context of that function. The type of `better_deco` is `partial[T@deco]` — i.e. it is parameterized by a type variable that is now out of scope. When you then apply `better_deco` as a decorator, this out-of-scope type variable can no longer be "solved".

I'm not sure why this works in mypy. It seems like it should not given the scoping rules defined in PEP 484. It's possible that it's a bug, and mypy is not enforcing these scoping rules in some code path. Or maybe it makes a special exception in certain cases, although this could lead to incorrect behavior in other cases. It's also possible that my understanding of the scoping rules is incomplete. (PEP 484 lacks clarity on some of the finer details of type variable scoping.)

The workaround that you discovered is the way that I would suggest handling this case.

**Comment by erictraut** [comment]
I'm going to close the issue because I think pyright is working as expected in this case.
